hera_mc/cm_partconnect.py

```diff
@@ -47,38 +47,60 @@ def __repr__(self):
                 '{self.stop_gpstime}>'.format(self=self))
 
 
-def add_part_rosetta(session, hpn, syspn, start_date, stop_date=None):
+def update_part_rosetta(hpn, syspn, at_date, date2=None, session=None):
     """
-    Add part information into database.
+    Update rosetta relationship in part_rosetta table.
+
+    If hpn <-> syspn relationship is active, will close it at at_date.
+    If hpn <-> syspn relationship is NOT active, it will start it at at_date.
+    If date2 is None, at_date is used as either start or stop, as appropriate.
+    Note that hpn is cast to upper, but syspn is not modified.
 
     Parameters
     ----------
-    session : object
-        Database session to use.  If None, it will start a new session, then close.
     hpn : str
         HERA part number
     syspn : str
         System part number
-    start_date : any format that cm_utils.get_astropytime understands
-        Date to use for the start
-    stop_date : any format that cm_utils.get_astropytime understands
-        Date to use for the stop
+    at_date : any format that cm_utils.get_astropytime understands
+        Date to use for action: start for new is relationship not active, else stop.
+    date2 : any format that cm_utils.get_astropytime understands
+        If not None, use for stop
+    session : object
+        Database session to use.  If None, it will start a new session, then close.
     """
+    hpn = hpn.upper()
+    at_date = int(cm_utils.get_astropytime(at_date).gps)
+    if date2 is not None:
+        date2 = int(cm_utils.get_astropytime(date2).gps)
+
     close_session_when_done = False
     if session is None:  # pragma: no cover
         db = mc.connect_to_mc_db(None)
         session = db.sessionmaker()
         close_session_when_done = True
 
-    rose = PartRosetta()
-    rose.hpn = hpn
-    rose.syspn = syspn
-    rose.start_gpstime = int(cm_utils.get_astropytime(start_date).gps)
-    if stop_date is None:
-        rose.stop_gpstime = None
+    old_rose = None
+    ctr = 0
+    for trial in session.query(PartRosetta).filter(
+            (func.upper(PartRosetta.hpn) == hpn) &  # noqa
+            (func.upper(PartRosetta.syspn) == syspn.upper())):
+        if trial.stop_gpstime is None:
+            ctr += 1
+            old_rose = trial
+    if ctr > 1:
+        raise ValueError("Multiple rosetta relationship active for {} - {}".format(hpn, syspn))
+    if old_rose is None:
+        new_rose = PartRosetta()
+        new_rose.hpn = hpn
+        new_rose.syspn = syspn
+        new_rose.start_gpstime = at_date
+        new_rose.stop_gpstime = date2
+        session.add(new_rose)
```

Comment:
This seems potentially dangerous to me, in that it seems that if I ask for a pairing that doesn't exist it makes a new one without checking to see if there is an existing one for that hpn already. So it can generate multiple relationships between an hpn and different syspns, for example. Its that desired?

Reply:
I had left open the possibility of more than one syspn active, but in retrospect that is probably not desirable.  One may have more than one syspn per hpn though.  So, I removed the filter on hpn and added a warning if trying to add a stop date where one exists.  Also updated checking for active relation.

Reply:
What's the use case for more that one syspn per hpn? I'm thinking about this through the filter of SNAP hostnames. I never want more than one SNAP hostname per physical SNAP I don't think.

Reply:
Currently, many snaps have multiple hostnames.  This is mostly historical I think, but given this fact I kept open the possibility.

Reply:
ah, ok.